src/plugins/syscalls/win.cpp

```diff
@@ -202,6 +202,87 @@ static std::vector<uint64_t> extract_args(drakvuf_t drakvuf, drakvuf_trap_info_t
     return args;
 }
 
+static bool resolve_dll(drakvuf_t drakvuf, drakvuf_trap_info_t* info, const char* dllname, addr_t* base, addr_t* size)
+{
+    resolve_ctx_t ctx{ .name = dllname };
+
+    drakvuf_enumerate_process_modules(drakvuf, info->proc_data.base_addr,
+        [](drakvuf_t dravkuf, const module_info_t* module_info, bool* need_free, bool* need_stop, void* ctx)
+    {
+        auto c = static_cast<resolve_ctx_t*>(ctx);
+
+        if (!strcmp((const char*)module_info->base_name->contents, c->name))
+        {
+            c->base    = module_info->base_addr;
+            c->size    = module_info->size;
+            *need_stop = true;
+        }
+        return true;
+    }, &ctx);
+
+    if (ctx.size && ctx.base)
+    {
+        *base = ctx.base;
+        *size = ctx.size;
+        return true;
+    }
+    return false;
+}
+
+static bool is_inlined_syscall(drakvuf_t drakvuf, drakvuf_trap_info_t* info, syscalls* s, const char* subsystem)
+{
+    // Only x64 nt syscalls are supported.
+    if (s->is32bit || !s->kernel_size || strcmp(subsystem, "nt"))
+        return false;
+
+    const addr_t rspbase = drakvuf_get_rspbase(drakvuf, info);
+    const addr_t diff    = rspbase - info->regs->rsp;
+    // Here we check if the call originated from usermode (syscall) or from other driver (iat call).
+    // KiSystemCall64 allocates 0x158 + 7 * 8 = 0x190 bytes. The qword at offset 0x28 - return address to usermode:
+    // -0x00:  mov     rsp, gs:1A8h
+    // -0x08:  push    2Bh
+    // -0x10:  push    qword ptr gs:10h
+    // -0x18:  push    r11
+    // -0x20:  push    33h
+    // -0x28:  push    rcx
+    // -0x28:  mov     rcx, r10
+    // -0x30:  sub     rsp, 8
+    // -0x38:  push    rbp
+    // -0x190: sub     rsp, 158h
+    // Since we hook Nt* functions, there are situations when the call originated from other driver and not from usermode.
+    // This checks if stack displacement is less than 0x190 + 8 (call instruction) + N (number of function arguments that are pushed on the stack).
+    if (0x190 > diff || diff > 0x220)
```

Comment:
This seems to be very fragile with no guarantees it won't break with a different release if the kernel.

Reply:
It was tested on windows7 up to the latest windows 11, the KiSyscall64 prologue is always the same.
```
    // -0x00:  mov     rsp, gs:1A8h
    // -0x08:  push    2Bh
    // -0x10:  push    qword ptr gs:10h
    // -0x18:  push    r11
    // -0x20:  push    33h
    // -0x28:  push    rcx
```

Reply:
Could you at least make the comparison ops the same order? Ie. the variable always on the left or the right of the comparison.

Reply:
sure!